src/main/java/io/cdap/plugin/sap/OData2Client.java

```diff
@@ -0,0 +1,170 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.sap;
+
+import com.google.common.base.Strings;
+import io.cdap.plugin.sap.exception.ODataException;
+import org.apache.olingo.odata2.api.edm.Edm;
+import org.apache.olingo.odata2.api.edm.EdmEntitySet;
+import org.apache.olingo.odata2.api.edm.EdmEntityType;
+import org.apache.olingo.odata2.api.edm.EdmException;
+import org.apache.olingo.odata2.api.ep.EntityProvider;
+import org.apache.olingo.odata2.api.ep.EntityProviderException;
+import org.apache.olingo.odata2.api.ep.EntityProviderReadProperties;
+import org.apache.olingo.odata2.api.ep.feed.ODataFeed;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.List;
+import javax.annotation.Nullable;
+import javax.ws.rs.HttpMethod;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+
+/**
+ * Provides handy methods to consume OData v2 services using Apache Olingo v2 as Client Library.
+ */
+public class OData2Client {
+
+  private static final String METADATA = "$metadata";
+  private static final String SEPARATOR = "/";
+  private static final String QUERY_SEPARATOR = "?";
+
+  private String rootUrl;
+  private String username;
+  private String password;
+
+  /**
+   * @param rootUrl  URL of the OData service. The URL must end with an external service name
+   *                 (e.g., http://eccsvrname:8000/sap/opu/odata/sap/zgw100_dd02l_so_srv/).
+   * @param username username for basic authentication.
+   * @param password password for basic authentication.
+   */
+  public OData2Client(String rootUrl, String username, String password) {
+    this.rootUrl = rootUrl;
+    this.username = username;
+    this.password = password;
+  }
+
+  /**
+   * Query {@link ODataFeed} using the specified entity set name and OData query.
+   *
+   * @param entitySetName entity set name.
+   * @param query         query such as "$top=2&$select=BuyerName&$filter=BuyerName eq 'TECUM'".
+   * @return {@link ODataFeed} for the specified entity set name.
+   * @throws ODataException if the specified entity set cannot be read.
+   */
+  public ODataFeed queryEntitySet(String entitySetName, @Nullable String query) {
+    Edm metadata = getMetadata();
+
+    try {
+      EdmEntitySet entitySet = metadata.getDefaultEntityContainer().getEntitySet(entitySetName);
+      String entitySetUrl = Strings.isNullOrEmpty(query) ?
+        rootUrl + SEPARATOR + entitySetName : rootUrl + SEPARATOR + entitySetName + QUERY_SEPARATOR + query;
```

Comment:
Instead of having this duplicate concatenation, it's a little easier to read if you have a variable for rootUrl + SEPARATOR + entitySetName, then append to it if query is not null or empty.

Reply:
also, is it ever validated that the query is well formed? For example, are we sure it doesn't start with a '?'

Reply:
Changed to use a variable. Added query validation logic and unit tests.